leopard.go

```diff
@@ -0,0 +1,846 @@
+	// I tried rolling the memcpy/memset into the first layer of the FFT and
+	// found that it only yields a 4% performance improvement, which is not
+	// worth the extra complexity.
+	for i := 0; i < mtrunc; i++ {
+		copy(work[i], data[i])
+	}
+	for i := mtrunc; i < m; i++ {
+		memset(work[i], 0)
+	}
+
+	// I tried splitting up the first few layers into L3-cache sized blocks but
+	// found that it only provides about 5% performance boost, which is not
+	// worth the extra complexity.
+
+	// Decimation in time: Unroll 2 layers at a time
+	dist := 1
+	dist4 := 4
+	for dist4 <= m {
+		// For each set of dist*4 elements:
+		for r := 0; r < mtrunc; r += dist4 {
+			iend := r + dist
+			log_m01 := skewLUT[iend]
+			log_m02 := skewLUT[iend+dist]
+			log_m23 := skewLUT[iend+dist*2]
+
+			// For each set of dist elements:
+			for i := r; i < iend; i++ {
+				ifftDIT4(
+					work[i:],
+					dist,
+					log_m01,
+					log_m23,
+					log_m02)
+			}
+		}
+
+		dist = dist4
+		dist4 <<= 2
+		// I tried alternating sweeps left->right and right->left to reduce cache misses.
+		// It provides about 1% performance boost when done for both FFT and IFFT, so it
+		// does not seem to be worth the extra complexity.
+	}
+
+	// If there is one layer left:
+	if dist < m {
+		// Assuming that dist = m / 2
+		if dist*2 != m {
+			panic("internal error")
+		}
+
+		logm := skewLUT[dist]
+
+		if logm == modulus {
+			xorSlices(work[dist:dist*2], work[:dist])
+		} else {
+			for i := 0; i < dist; i++ {
+				ifftDIT2(work[i], work[i+dist], logm)
+			}
+		}
+	}
+
+	// I tried unrolling this but it does not provide more than 5% performance
+	// improvement for 16-bit finite fields, so it's not worth the complexity.
+	if xorRes != nil {
+		xorSlices(xorRes[:m], work[:m])
+	}
+}
+
+// 4-way butterfly
+func ifftDIT4(work [][]byte, dist int, log_m01, log_m23, log_m02 ffe) {
+	// First layer:
+	if log_m01 == modulus {
+		xor(work[dist], work[0])
+	} else {
+		ifftDIT2(work[0], work[dist], log_m01)
+	}
+
+	if log_m23 == modulus {
+		xor(work[dist*3], work[dist*2])
+	} else {
+		ifftDIT2(work[dist*2], work[dist*3], log_m23)
+	}
+
+	// Second layer:
+	if log_m02 == modulus {
+		xor(work[dist*2], work[0])
+		xor(work[dist*3], work[dist])
+	} else {
+		ifftDIT2(work[0], work[dist*2], log_m02)
+		ifftDIT2(work[dist], work[dist*3], log_m02)
+	}
+}
+
+// 2-way butterfly
+func ifftDIT2(x, y []byte, log_m ffe) {
+	// Reference version:
+	xor(y, x)
+	refMulAdd(x, y, log_m)
+}
+
+// Reference version of muladd: x[] ^= y[] * log_m
+func refMulAdd(x, y []byte, log_m ffe) {
+	lut := mul16LUTs[log_m].LUT
+
+	for off := 0; off < len(x); off += 64 {
+		for i := 0; i < 32; i++ {
+			lo := y[off+i]
+			hi := y[off+i+32]
+
+			prod :=
+				lut[(lo&15)] ^
+					lut[(lo>>4)+16] ^
+					lut[(hi&15)+32] ^
+					lut[(hi>>4)+48]
+
+			x[off+i] ^= byte(prod)
+			x[off+i+32] ^= byte(prod >> 8)
+		}
+	}
+}
+
+func memset(s []byte, n byte) {
+	for i := range s {
+		s[i] = 0
```

Comment:
You want to set to this to n, but given that we invoke memset(s, 0) perhaps we can ensure that the compiler can recognize memset indeed as clearing a slice by making a specialized function name say: "memset0(s)"

Reply:
Good catch and good point. I've specialized `memset` to `memclr` to encourage compiler optimization.